### DISCUSSION TITLE
How to list all allowed paths for direnv

### DISCUSSION THREAD

**Discussion by avijitpal9** [direnv/direnv#1092] [Q&A]
Hi All,
I have recently started using direnv. I am wondering is there a way to list all allowed path for direnv. I cannot find any direnv command called list or ls which would provide that information. direnv status command only list .envrc for current directory if present.

**Comment by lucascool12** [comment]
According to the [man page](https://github.com/direnv/direnv/blob/master/man/direnv.1.md#files): `$XDG_DATA_HOME/direnv/allow` records the `.envrc` files that have been allowed.
The contents of the files in this directory contain the paths to the `.envrc` file they allow.
These files seems to have a hash of the contents of the `.envrc` file as name. Meaning different files can contain the same path.
Using `ls ${XDG_DATA_HOME:-$HOME/.local}/share/direnv/allow/* -d` we can print all the absolute paths of all the files in
`$XDG_DATA_HOME/direnv/allow`. 
We can then pipe this to xargs with cat as command to print the list of paths that have been allowed: 
`ls ${XDG_DATA_HOME:-$HOME/.local}/share/direnv/allow/* -d | xargs cat`.
This works but might contain duplicates. We can use `uniq` to filter these duplicates. `uniq` removes repeating lines so this requires us to first sort the output.

So finally here is the command that lists all the allowed paths:
`ls ${XDG_DATA_HOME:-$HOME/.local}/share/direnv/allow/* -d | xargs cat | sort | uniq`
You can also create a python script or something else to do the exact same with the info supplied above.

**Reply by lucascool12** [reply]
When I thought about this solution again I noticed I completely forgot about `.envrc` files that have been allowed but no longer exist and where the `.envrc` file has previously been allowed but now has different contents.
Here is a shell script that does these things: check if the file actually exists and if the hash of the file exists inside the direnv allow directory.
```sh
#!/bin/sh

allow_dir=${XDG_DATA_HOME:-$HOME/.local}/share/direnv/allow
allow_paths=`ls $allow_dir/* -d | xargs cat | sort | uniq`
for allow_path in $allow_paths; do
    if [[ ! -e $allow_path ]]; then
        continue
    fi
    hash=`printf "$allow_path\n" | cat - $allow_path | sha256sum - | cut -d ' ' -f 1`
    if [[ `find $allow_dir -name $hash` ]]; then
        echo $allow_path
    fi
done
```
The file names in the direnv allow directory is a sha256 hash of the path to the `.envrc` file + newline and then the contents of the file, as you can see in this [function](https://github.com/direnv/direnv/blob/176bafc44922bd791ce95ed771a88cd52f8f9f70/internal/cmd/rc.go#L263), hence the:
```sh
`printf "$allow_path\n" | cat - $allow_path | sha256sum - | cut -d ' ' -f 1`
``` 


Hopefully I don't make another reply explaining why my previous answer was not correct tomorrow.
